File: CopyPaste.Core/AppLogger.cs
using System.Collections.Concurrent;
using System.Globalization;
using System.Runtime.CompilerServices;
using System.Text;

namespace CopyPaste.Core;

/// <summary>
/// Simple file logger optimized for AOT and minimal allocations.
/// Logs are written to the application data folder.
/// </summary>
public static class AppLogger
{
    private static readonly string _logDirectory = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "CopyPaste", "logs");

    private static readonly string _logFilePath = Path.Combine(
        _logDirectory,
        $"copypaste_{DateTime.Now:yyyy-MM-dd}.log");

    private static readonly ConcurrentQueue<string> _logQueue = new();
    private static readonly Lock _writeLock = new();
    private static bool _isInitialized;
    private static bool _isEnabled = true;

    private const int _maxLogAgeDays = 7;
    private const int _maxLogSizeMb = 10;

    /// <summary>
    /// Enable or disable logging at runtime.
    /// </summary>
    public static bool IsEnabled
    {
        get => _isEnabled;
        set => _isEnabled = value;
    }

    /// <summary>
    /// Initialize the logger and clean old log files.
    /// </summary>
    public static void Initialize()
    {
        if (_isInitialized) return;

        try
        {
            Directory.CreateDirectory(_logDirectory);
            CleanOldLogs();
            _isInitialized = true;

            Info("Logger initialized", "AppLogger");
        }
        catch
        {
            _isEnabled = false;
        }
    }

    /// <summary>
    /// Log an informational message.
    /// </summary>
    public static void Info(string message, [CallerMemberName] string caller = "") =>
        Log("INFO", message, caller);

    /// <summary>
    /// Log a warning message.
    /// </summary>
    public static void Warn(string message, [CallerMemberName] string caller = "") =>
        Log("WARN", message, caller);

    /// <summary>
    /// Log an error message.
    /// </summary>
    public static void Error(string message, [CallerMemberName] string caller = "") =>
        Log("ERROR", message, caller);

    /// <summary>
    /// Log an exception with full details.
    /// </summary>
    public static void Exception(Exception? ex, string context = "", [CallerMemberName] string caller = "")
    {
        if (!_isEnabled || ex is null) return;

        var sb = new StringBuilder();
        sb.Append(context);
        if (!string.IsNullOrEmpty(context)) sb.Append(" - ");
        sb.Append(ex.GetType().Name);
        sb.Append(": ");
        sb.Append(ex.Message);

        if (ex.StackTrace is not null)
        {
            sb.AppendLine();
            sb.Append("  StackTrace: ");
            sb.Append(ex.StackTrace);
        }

        if (ex.InnerException is not null)
        {
            sb.AppendLine();
            sb.Append("  Inner: ");
            sb.Append(ex.InnerException.Message);
        }

        Log("ERROR", sb.ToString(), caller);
    }

    /// <summary>
    /// Gets the path to the current log file.
    /// </summary>
    public static string LogFilePath => _logFilePath;

    /// <summary>
    /// Gets the logs directory path.
    /// </summary>
    public static string LogDirectory => _logDirectory;

    private static void Log(string level, string message, string caller)
    {
        if (!_isEnabled || !_isInitialized) return;

        var timestamp = DateTime.Now.ToString("HH:mm:ss.fff", CultureInfo.InvariantCulture);
        var logEntry = $"[{timestamp}] [{level}] [{caller}] {message}";

        _logQueue.Enqueue(logEntry);
        FlushQueue();
    }

    private static void FlushQueue()
    {
        if (!_logQueue.TryPeek(out _)) return;

        lock (_writeLock)
        {
            try
            {
                // Check file size before writing
                if (File.Exists(_logFilePath))
                {
                    var fileInfo = new FileInfo(_logFilePath);
                    if (fileInfo.Length > _maxLogSizeMb * 1024 * 1024)
                    {
                        RotateLog();
                    }
                }

                using var writer = new StreamWriter(_logFilePath, append: true, Encoding.UTF8);
                while (_logQueue.TryDequeue(out var entry))
                {
                    writer.WriteLine(entry);
                }
            }
            catch
            {
                // Silently fail - logging should never crash the app
            }
        }
    }

    private static void RotateLog()
    {
        var rotatedPath = Path.Combine(_logDirectory,
            $"copypaste_{DateTime.Now:yyyy-MM-dd_HHmmss}.log");

        try
        {
            File.Move(_logFilePath, rotatedPath);
        }
        catch
        {
            // If rotation fails, just delete the old log
            File.Delete(_logFilePath);
        }
    }

    private static void CleanOldLogs()
    {
        try
        {
            var cutoffDate = DateTime.Now.AddDays(-_maxLogAgeDays);
            var logFiles = Directory.GetFiles(_logDirectory, "copypaste_*.log");

            foreach (var file in logFiles)
            {
                var fileInfo = new FileInfo(file);
                if (fileInfo.LastWriteTime < cutoffDate)
                {
                    File.Delete(file);
                }
            }
        }
        catch
        {
            // Ignore cleanup errors
        }
    }
}

File: v2/core/lib/services/app_logger.dart
import 'dart:io';

import 'package:path/path.dart' as p;

class AppLogger {
  AppLogger._();

  static String? _logDirectory;
  static String? _logFilePath;
  static bool _isInitialized = false;
  static bool isEnabled = true;

  static const int _maxLogAgeDays = 7;
  static const int _maxLogSizeBytes = 10 * 1024 * 1024;

  static String? get logFilePath => _logFilePath;
  static String? get logDirectory => _logDirectory;

  static void initialize(String logsPath) {
    if (_isInitialized) return;
    try {
      _logDirectory = logsPath;
      final now = DateTime.now();
      final dateStr =
          '${now.year}-${_pad(now.month)}-${_pad(now.day)}';
      _logFilePath = p.join(logsPath, 'copypaste_$dateStr.log');
      Directory(logsPath).createSync(recursive: true);
      _cleanOldLogs();
      _isInitialized = true;
      info('Logger initialized');
    } catch (_) {
      isEnabled = false;
    }
  }

  static void info(String message) => _log('INFO', message);

  static void warn(String message) => _log('WARN', message);

  static void error(String message) => _log('ERROR', message);

  static void exception(Object? error, [StackTrace? stackTrace, String context = '']) {
    if (!isEnabled || !_isInitialized || error == null) return;
    final sb = StringBuffer();
    if (context.isNotEmpty) sb.write('$context - ');
    sb.write(error.toString());
    if (stackTrace != null) sb.write('\n$stackTrace');
    _log('ERROR', sb.toString());
  }

  static void _log(String level, String message) {
    if (!isEnabled || !_isInitialized || _logFilePath == null) return;
    try {
      final now = DateTime.now();
      final timestamp =
          '${_pad(now.hour)}:${_pad(now.minute)}:${_pad(now.second)}.${_pad3(now.millisecond)}';
      final entry = '[$timestamp] [$level] $message\n';
      final file = File(_logFilePath!);
      if (file.existsSync() && file.lengthSync() > _maxLogSizeBytes) {
        _rotateLog();
      }
      file.writeAsStringSync(entry, mode: FileMode.append);
    } catch (_) {}
  }

  static void _rotateLog() {
    if (_logFilePath == null || _logDirectory == null) return;
    try {
      final now = DateTime.now();
      final rotatedName =
          'copypaste_${now.year}-${_pad(now.month)}-${_pad(now.day)}_${_pad(now.hour)}${_pad(now.minute)}${_pad(now.second)}.log';
      File(_logFilePath!).renameSync(p.join(_logDirectory!, rotatedName));
    } catch (_) {
      try {
        File(_logFilePath!).deleteSync();
      } catch (_) {}
    }
  }

  static void _cleanOldLogs() {
    if (_logDirectory == null) return;
    try {
      final cutoff = DateTime.now().subtract(
        const Duration(days: _maxLogAgeDays),
      );
      final dir = Directory(_logDirectory!);
      if (!dir.existsSync()) return;
      for (final file in dir.listSync().whereType<File>()) {
        if (p.basename(file.path).startsWith('copypaste_') &&
            file.path.endsWith('.log')) {
          if (file.lastModifiedSync().isBefore(cutoff)) {
            file.deleteSync();
          }
        }
      }
    } catch (_) {}
  }

  static String _pad(int n) => n.toString().padLeft(2, '0');
  static String _pad3(int n) => n.toString().padLeft(3, '0');
}
